### ISSUE TITLE
[Flang] Missing diagnose when procedure pointer is subroutine but the pointer assignment target is a function

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#83273]
Consider the following code:
```
    IMPLICIT NONE ! p is not implicitly typeable
    REAL, EXTERNAL :: func
    PROCEDURE(), POINTER :: p1

    p1 => func  ! illegal
    END PROGRAM
```
`p1` is not explicitly or implicitly typed, so it should represent a subroutine.

Flang currently compiles the above code successfully.

**Comment by klausler** [comment]
The code conforms and no error should be reported.  `PROCEDURE(), POINTER` does not declare a subroutine, it declares a procedure pointer that is not known to be either a subroutine or a function.  See 15.4.3.6 paragraph 4: "If proc-interface does not appear, the procedure declaration statement does not specify whether the declared procedure entities are subroutines or functions."

**Comment by DanielCChen** [comment]
Yeah, I read that wording. I think the question is how the implicit typing rule should apply. In this particular case, `implicit none` is present, and `p1` is not explicitly` typed before the pointer assignment. The only thing that `p1` can pointer assignment to is subroutine. Any function target wouldn't make sense because of the impactable return type, which `p1` has none. 

Flang issues an error for the case in Issue #82078. I copy it here too for reference.
```
Program main
  interface
    integer function foo()
    end
  end interface
  procedure(), pointer :: p1
  integer :: res
  p1 => foo
  res = p1()
end
```
The justification is that p1 is implicitly typed as `default real`. For the same reason, I think Flang should issue an error for the code in this issue.

**Comment by klausler** [comment]
In that second case, `p1` is called as a function, so it's a function.

If a procedure pointer has no interface, and is not referenced as a function or a subroutine, it is neither, and it doesn't require a type.  See also 10.2.2.4 paragraph 5.

**Comment by DanielCChen** [comment]
I see. It seems Flang checks it at the reference. If I add a reference to `p1` as
```
    IMPLICIT NONE ! p is not implicitly typeable
    REAL, EXTERNAL :: func
    PROCEDURE(), POINTER :: p1
    real :: res

    p1 => func  ! illegal
    res = p1()
    END PROGRAM
    ```
  Flang issues an error 
  ./t.f:3:29: error: No explicit type declared for 'p1'
      PROCEDURE(), POINTER :: p1
                              ^^
   ```
   
   This error should be issued for referencing `p1` if the target is of any function.

**Comment by klausler** [comment]
As I explained above, the reference to `p1` as a function in your modified test forces `p1` to be a function and to require a type.  In the original test case for this bug, there is no reference to the pointer as either a function or as a subroutine, and no error should appear.
